Integrate Elasticsearch Query Language, ES|QL

ChrisHegarty opened the pull request:
Integrate the Elasticsearch Query Language, ES|QL.

This Pull Request offers a consolidated view over the commits of the ES|QL feature branch, allowing to review and test them altogether, and in one place.

Upon integration into _main_ this PR will merge the commits as-is (not squashed). This will effectively preserve the history of the ES|QL code.

nik9000 committed:
Make most singleton aggs null on empty (ESQL-1300)

In SQL `AVG(foo)` is `null` if there are no value for `foo`. Same for
`MIN(foo)` and `MAX(foo)`. In fact, the only functions that don't return
`null` on empty inputs seem to be `COUNT` and `COUNT(DISTINCT`.

This flips our non-grouping aggs to have the same behavior because it's
both more expected and fits better with other things we're building.
This *is* different from Elasticsearch's aggs. But it's different in a
good way. It also lines up more closely with the way that our grouping
aggs work.

This also revives the broken `AggregatorBenchmark` so that I could get
performance figures for this change. And it's within the margin of
error:

```
    (blockType)  (grouping)  (op)  Mode  Cnt   Before         After         Units
   vector_longs        none   sum  avgt    7   0.440 ± 0.017  0.397 ± 0.003 ns/op
half_null_longs        none   sum  avgt    7   5.785 ± 0.022  5.861 ± 0.134 ns/op
```

I expected a small slowdown on the `half_null_longs` line and see it,
but is within the margin of error. Either way, that's not the line
that's nearly as optimized. We'll loop back around to it eventually.

Closes ESQL-1297

elasticsearchmachine committed:
Merge pull request ESQL-1301 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1305 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Emit warnings from bad `date_parse` (ESQL-1303)

This rigs the exceptions caught by `warnExceptions` in `Evaluator` to
emit warnings through Elasticsearch's warnings system, the same way that
the conversion functions work.

Closes ESQL-1211

costin committed:
Introduce SurrogateExpression (ESQL-1285)

Convert Avg into a SurrogateExpression and introduce dedicated rule
 for handling surrogate AggregateFunction
Remove Avg implementation
Use sum instead of avg in some planning test
Add dataType case for Div operator

Relates ESQL-747

nik9000 committed:
Create e() function (ESQL-1304)

Euler's number.

nik9000 committed:
Enable `min` and `max` aggs on dates (ESQL-1302)

This enables the `min` and `max` aggs on `date` fields without enabling
any of the other numeric aggregates on `date`s. It also adds a fairly
paranoid test that `sum` is not enabled on `date`s because that doesn't
make a whole lot of sense.

Closes ESQL-1247

elasticsearchmachine committed:
Merge pull request ESQL-1309 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1310 from elastic/main

🤖 ESQL: Merge upstream

dnhatn committed:
Support multi values in enrich lookup (ESQL-1298)

This PR adds a new method that allows appending a block to a builder at
a single position. This method is required to support multi-values in
the enrich lookup. Ideally, the new method should be unified with the
`copyFrom` method, but I will address it as a follow-up to reduce the
complexity of this PR.

Closes ESQL-1280

elasticsearchmachine committed:
Merge pull request ESQL-1312 from elastic/main

🤖 ESQL: Merge upstream

luigidellaquila committed:
Add docs for ENRICH command (ESQL-1313)


Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>

costin committed:
Make Median a surrogate expression (ESQL-1307)

luigidellaquila committed:
Add query parameters to ESQL (ESQL-1308)

elasticsearchmachine committed:
Merge pull request ESQL-1314 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1318 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1319 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1320 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Ensure grouping agg behavior on only null fields (ESQL-1311)

This adds an explicit test that makes sure that grouping aggs (`STATS
FOO(x) BY y`) *mostly* return `null` if they receive only null values.
We have code for this scattered around the grouping aggs but we were
only testing it *sometimes*. This tests it all the time. Also! The
behavior wasn't *quite* consistent. `COUNT` and `COUNT(DISTINCT` style
aggs should return `0` but they didn't all do that. And non-`COUNT`
style aggs should return `null` and they all did that. Exception `SUM`
on doubles.

abdonpijpelink committed:
[DOCS] Clarify the order for RENAME

abdonpijpelink committed:
Remove unnecessary space

elasticsearchmachine committed:
Merge pull request ESQL-1323 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1325 from elastic/main

🤖 ESQL: Merge upstream

costin committed:
Pick changes upstream

luigidellaquila committed:
Implement now() function (ESQL-1172)

returns current datetime

nik9000 committed:
Implement `MV_DEDUPE` (ESQL-1287)

This implements the `MV_DEDUPE` function that removes duplicates from
multivalues fields. It wasn't strictly in our list of things we need in
the first release, but I'm grabbing this now because I realized I needed
very similar infrastructure when I was trying to build grouping by
multivalued fields. In fact, I realized that I could use our
stringtemplate code generation to generate most of the complex parts.
This generates the actual body of `MV_DEDUPE`'s implementation and the
body of the `Block` accepting `BlockHash` implementations. It'll be
useful in the final step for grouping by multivalued fields.

I also got pretty curious about whether the `O(n^2)` or `O(n*log(n))`
algorithm for deduplication is faster. I'd been assuming that for all
reasonable sized inputs the `O(n^2)` bubble sort looking selection
algorithm was faster. So I measured it. And it's mostly true - even for
`BytesRef` if you have a dozen entries the selection algorithm is
faster. Lower overhead and stuff. Anyway, to measure it I had to
implement the copy-and-sort `O(n*log(n))` algorithm. So while I was
there I plugged it in and selected it in cases where the number of
inputs is large and the selection alogorithm is likely to be slower.

abdonpijpelink committed:
Merge pull request ESQL-1322 from abdonpijpelink/rename-clarify-order

[DOCS] Clarify the field order for RENAME

elasticsearchmachine committed:
Merge pull request ESQL-1330 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Fix compilation

What happened?

elasticsearchmachine committed:
Merge pull request ESQL-1332 from elastic/main

🤖 ESQL: Merge upstream

costin committed:
Remove ESQL IsNull in favor of QL (ESQL-1329)

Move related optimization rules into QL

elasticsearchmachine committed:
Merge pull request ESQL-1337 from elastic/main

🤖 ESQL: Merge upstream

dnhatn committed:
Lookup enrich queries in input order (ESQL-1324)

We currently have the enrich lookup implemented in sorted mode. However,
we should have started with an unsorted mode, which performs the lookup
in the order of the input queries, for these reasons:

- Enrich indices are typically small, containing fewer than 10K
documents. With such small indices, the unsorted mode should outperform
the sorted mode as it eliminates the need for three sorting operations.

- The range match type does not benefit much from sorting queries.

This pull request removes the sorted mode and replaces it with the
unsorted mode. Also, this PR adds support for the range match type for
fields such as ip_range, date_range, and long_range. After the feature
freeze, I will reintroduce the sorted mode as an optimization
specifically for large enrich indices.

elasticsearchmachine committed:
Merge pull request ESQL-1341 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1343 from elastic/main

🤖 ESQL: Merge upstream

ChrisHegarty committed:
Towards multiple channel agg intermediate state (ESQL-1321)

This change makes the agg intermediate state explicit, and requires each
agg to declare it.

The local execution planner now determines the agg intermediate state.

The `AggregatorFunction` and `GroupingAggregatorFunction` interfaces get
a method that exposes the number of blocks used by the intermediate
state.
```
  /** The number of blocks used by intermediate state. */
  int intermediateBlockCount();
```

Each concrete agg implementation has to explicitly declare its
intermediate states e.g.

``` 
@Aggregator({
        @IntermediateState(name = "value", type = "DOUBLE"),
        @IntermediateState(name = "delta", type = "DOUBLE"),
        @IntermediateState(name = "nonnull", type = "BOOLEAN") })
@GroupingAggregator
class SumDoubleAggregator { .. }
```

The grouping and non-grouping states can be different, but if not the
grouping will defer to the non-grouping - thus we only need to declare
it once in many cases.

bpintea committed:
Add metadata type and function support (ESQL-1315)

This adds support for metadata fields `_index` and `_version`. These are
exported through a (newly added) function, `metadata()`; ex. `| eval
_index = metadata("_index")`.

The function's resolution checks if the defined field is supported. It's
then substituted with a newly introduced attribute, `MetadataAttribute`.
Some optimiser rules have been changed to allow the `Eval`s be pushed
down past the `Limit`, so that these are part of the fragments
dispatched to the data nodes.
The extraction is currently done through the same mechanism as the
`FieldAttribute`s.
The fields aren't accessible "directly" (i.e. `from idx | where _index >
0`) and aren't projected implicitly (i.e. just by `from index`).

ChrisHegarty committed:
Allow for randomisation in testSingleValueSparse (ESQL-1344)

Update asserts in _testSingleValueSparseXXX_ to allow for the
possibility, albeit rare, that the randomisation will create dense
blocks, e.g. small block with say 2 positions can easily be dense,
though this is quite rare.

ChrisHegarty committed:
Add Vector views over BigArray backed data (ESQL-1345)

This commit adds Boolean, Int, Long, and Double Vector views over
BigArray backed data.

The Vectors are Releasable, so one must be careful to release them.

elasticsearchmachine committed:
Merge pull request ESQL-1347 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1350 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1351 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1352 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Support multivalued fields on arbitrary grouping (ESQL-1340)

This adds support for multivalued fields the `PackedValuesHash` it does
so by encoding batches of values for each column into bytes and then
reading converted values row-wise. This works while also not causing
per-row megamorphic calls. There are megamorphic calls when the batch is
used up, but that should only hit a few times per block.

dnhatn committed:
Do not allow partial results in ESQL (ESQL-1349)

If some target shards are unavailable during the planning, ESQL may
return partial results without any indication. While we plan to support
retrying and handling partial results in the future, it is currently
most appropriate to fail them.

elasticsearchmachine committed:
Merge pull request ESQL-1355 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1362 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1363 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1364 from elastic/main

🤖 ESQL: Merge upstream

bpintea committed:
Add unsigned_long type support (ESQL-1289)

This adds support for the `unsigned_long` type.
The type can be now used with the defined math function, both scalar and
MV'ed, arithmetic and binary comparison ones.
The `to_unsigned_long()` conversion function is also added.

elasticsearchmachine committed:
Merge pull request ESQL-1365 from elastic/main

🤖 ESQL: Merge upstream

breskeby committed:
Fix deprecated api usage in gradle build scripts

breskeby committed:
Merge pull request ESQL-1366 from elastic/fix-gradle-deprecations

Fix deprecated api usage in gradle build scripts

dnhatn committed:
Use TDigest fast mode in ESQL (ESQL-1361)

Ideally, we should allow users to select the T-digest mode: fast or
accuracy; but the default should be the fast mode.

elasticsearchmachine committed:
Merge pull request ESQL-1369 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Add `PI` and `TAU` functions (ESQL-1357)

Adds functions for the constants `PI` and it's big brother `TAU`.

elasticsearchmachine committed:
Merge pull request ESQL-1371 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1372 from elastic/main

🤖 ESQL: Merge upstream

csoulios committed:
Add support for text fields (ESQL-1354)

This PR adds support for loading `text` fields which can not be stored
as doc_values.

If text values are stored in the index (set `"store": true` in the field
mapping) the text value will be loaded directly from the index.
Otherwise, the value will be loaded by extracting the text from _source.

nik9000 committed:
Clean `UnaryScalarFunction` slightly (ESQL-1359)

Our base class for `UnaryScalarFunction` only takes one argument because
it's, well, unary. But it was reporting type errors on that argument as
though it were the first of many. That's silly.

I also added some tests for the `Abs` function which extends our
`UnaryScalarFunction` that would have caught this error. While I was
there I ported `Length` from QL's `UnaryScalarFunction` to ours. Let's
use our stuff. Even if it's wrong we can change it without bothing QL.

Finally I added some javadocs and removed some unused code.

ChrisHegarty committed:
Replace aggstate in aggs whose internal state is one of the primitives (ESQL-1375)

This commit moves aggs whose internal state is one of the primitives,
over to the new intermediate agg state mechanism.

We can then remove quite a bit of the internal serialization logic.

elasticsearchmachine committed:
Merge pull request ESQL-1386 from elastic/main

🤖 ESQL: Merge upstream

not-napoleon committed:
Log base 10 for ESQL (ESQL-1358)

Introduces a unary scalar function for base 10 log, which is a thin
wrapper over the Java.Math implementation

---------

Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>

nik9000 committed:
Add `last_updated` to status (ESQL-1387)

Adds a `last_updated` time to the status reported by a diver. We only
report driver status after each batch and this can show us if the batch
is super behind somehow.

elasticsearchmachine committed:
Merge pull request ESQL-1392 from elastic/main

🤖 ESQL: Merge upstream

costin committed:
Pick breaking changes from upstream

abdonpijpelink committed:
Change ESQL into ES|QL

abdonpijpelink committed:
Remove meta tags

abdonpijpelink committed:
Move semicolons

abdonpijpelink committed:
Move IS_NULL, POW, ROUND, STARTS_WITH, SUBSTRING code snippets to CSV files

dnhatn committed:
Harden remote exchange flow (ESQL-1339)

Today, we initiate the computation on a remote node by concurrently
sending the data-node request and fetch-page requests. The aims to
minimize latency of ESQL requests by avoiding round trips. However, this
approach proves to be overly complex and heroic. It requires us to
handle pending fetch-page requests that arrive before the data-node
request. Additionally, we also need to leave the already-completed sink
on the remote node for some time to serve on-the-fly fetch-page
requests; otherwise, the requests won't be returned until they time out.

To address these issues, this pull request proposes a simpler flow. For
each computation on a remote node, we will first send a request to open
an exchange. Upon receiving the response, we will concurrently send both
the data-node requests and fetch-pages. This approach eliminates the
above problems.

Furthermore, I have added a disruption test (see EsqlDisruptionIT.java)
to ensure that ESQL doesn't produce unexpected results or resource leaks
when the cluster experiences issues, such as a network partition.

abdonpijpelink committed:
Merge remote-tracking branch 'upstream/esql/lang' into es-pipe-ql

ChrisHegarty committed:
Fix another few usages of CURRENT

abdonpijpelink committed:
Merge remote-tracking branch 'upstream/esql/lang' into es-pipe-ql

ChrisHegarty committed:
Minor changes to string template plugin to simplify upstream merging (ESQL-1394)

martijnvg committed:
Add trim function

This change adds a string `trim` function.

martijnvg committed:
made test a little bit more evil

martijnvg committed:
use ROW in docs and added test with ROWS

martijnvg committed:
Make the trim function efficient by performing trim on BytesRef instead of converting to a string and invoking trim and then converting back to BytesRef.

nik9000 committed:
Start to prevent massive grouping blocks (ESQL-1370)

When you group by more than one multivalued field we generate one ord
per unique tuple of value of from each column. So if you group by
```
a=(1, 2, 3) b=(2, 3) c=(4, 5, 5)
```

Then you get these grouping keys:
```
1, 2, 4
1, 2, 5
1, 3, 4
1, 3, 5
2, 2, 4
2, 2, 5
2, 3, 4
2, 3, 5
3, 2, 4
3, 3, 5
```

That's as many grouping keys the the product of the Set-wise cardinality
of all elements. "Product" is a dangerous word! It's possible to make a
simple document containing just two fields that each are a list of
10,000 values and then send *that* into the aggregation framework. That
little baby document will spit out 100,000,000 grouping ordinals!

Without this PR we'd try to create a single `Block` that contains that
many entries. Or, rather, it'd be as big as the nearest power of two.
Gigantonormous. About 760mb! Like, possible, but a huge "slug" of heap
usage and not great.

This PR changes it so, at least for pairs of `long` keys we'll make many
smaller blocks. We cut the emitted ordinals into a block no more than
16*1024 entries, the default length of a block. That means our baby
document would make 6103 full blocks and one half full block. But each
one is going less than 200kb.

Relates to ESQL-1360

costin committed:
Merge commit '6d45c57b8fc6d504fdc28f355c6db22ac811e629' into esql/lang

dnhatn committed:
Fix testRefreshSearchIdleShards (ESQL-1399)

This PR reworks `testRefreshSearchIdleShards` to disable shard
relocations and ensures that all target shards have pending refresh when
the query is being executed.

Closes ESQL-623

dnhatn committed:
Fix count_distinct intermediate state (ESQL-1400)

The `combineStates` of count_distinct uses an incorrect right position.
Unfortunately, our tests are not randomized enough to catch this bug.

elasticsearchmachine committed:
Merge pull request ESQL-1401 from elastic/main

🤖 ESQL: Merge upstream

martijnvg committed:
Reused example from spec file

abdonpijpelink committed:
One more set of meta attributes

martijnvg committed:
Change Trim#process to take one parameter that can be modified.

martijnvg committed:
spotless

abdonpijpelink committed:
Merge remote-tracking branch 'upstream/esql/lang' into es-pipe-ql

martijnvg committed:
Revert "spotless"

This reverts commit 97f4f2c8e51316f6d34727776cdf69fc9187cabb.

martijnvg committed:
Revert "Change Trim#process to take one parameter that can be modified."

This reverts commit 6b7c16d59e8f7138a41081a426ed0ec22b970836.

martijnvg committed:
iter

martijnvg committed:
Don't use provided scratch and always create a new BytesRef instance.

martijnvg committed:
Merge remote-tracking branch 'es/esql/lang' into string_trim

martijnvg committed:
Add trim function

This change adds a string trim function.

Relates to ESQL-1316

elasticsearchmachine committed:
Merge pull request ESQL-1407 from elastic/main

🤖 ESQL: Merge upstream

dnhatn committed:
Update threadpool changes in ESQL

elasticsearchmachine committed:
Merge pull request ESQL-1408 from elastic/main

🤖 ESQL: Merge upstream

abdonpijpelink committed:
Merge pull request ESQL-1393 from abdonpijpelink/es-pipe-ql

[DOCS] Change ESQL into ES|QL and other docs improvements

nik9000 committed:
Add status to `LIMIT` (ESQL-1388)

This adds a `Status` response to the `LIMIT` operator so we can monitor
them a little more closely.

dnhatn committed:
Make page_size parameter configurable (ESQL-1402)

Today, we have a hard-coded maximum page size of 16K in Lucene operators
and other operators like TopN and HashOperator. This default value
should work well in production. However, it doesn't provide enough
randomization in our tests because we mostly emit a single page.
Additionally, some tests take a significant amount of time because they
require indexing a large number of documents, which is several times the
page size.

To address these, this PR makes the page size parameter to be
configurable via the query pragmas, enabling randomization in tests.
This change has already uncovered a bug in LongLongBlockHash.

ChrisHegarty committed:
Refactor and remove aggstate (ESQL-1410)

This commits refactors the remaining intermediate state of aggregators
to use standard block types, and allows to remove the existing aggstate.

elasticsearchmachine committed:
Merge pull request ESQL-1418 from elastic/main

🤖 ESQL: Merge upstream

costin committed:
Add primitive variants in Log10 (ESQL-1409)

Check the type of the receiving field to pick the proper evaluator -
this is needed in order to do casting.

nik9000 committed:
Batch remaining grouping impls (ESQL-1406)

This batches all remaining `BlockHash` implementations that can produce
a combinatorial explosion of results. We don't bother batching any paths
that don't produce more results than the incoming page size though,
because we figure it's safe to keep working in batches of the size of
the page. This means that single element `BlockHash`es don't batch and
that multi-element `BlockHash`es never batch on their `Vector` branches,
if they have `Vector` branch.

dnhatn committed:
Mute testExtractFields

Tracked at ESQL-1420

dnhatn committed:
Remove outdated asserts in TopNOperatorTests

dnhatn committed:
Simplify state in Aggregator and Limit Operators (ESQL-1415)

Replace the state enum with the finished flag.

elasticsearchmachine committed:
Merge pull request ESQL-1421 from elastic/main

🤖 ESQL: Merge upstream

ChrisHegarty committed:
Remove Experimental from plan related types (ESQL-1417)

This commit removes the Experimental annotation from plan related types.
This part of the code is no longer Experimental.

ChrisHegarty committed:
Remove Experimental from operators (ESQL-1416)

This commit removes the Experimental annotation from operators.
This part of the code is no longer Experimental.

ChrisHegarty committed:
Remove Experimental from aggs (ESQL-1414)

This commit removes the Experimental annotation from aggs.
This part of the code is no longer Experimental.

ChrisHegarty committed:
Remove last vestiges of Experimental (ESQL-1425)

This commit removes the last vestiges of Experimental, as it is no
longer used.

dnhatn committed:
Reset Lucene scorer if executing thread changed (ESQL-1422)

With a very small page_size, Driver might take many iterations, then it
yields and to be re-executed in a different thread. This leads to the
below assertion, where Lucene strictly requires some instances such as
Scorer, DocValues must be accessed only by the thread that created them.
Although, this requirement is not enforced in production, we already
conform this in ValuesSourceReaderOperator and LuceneTopN. However, we
miss the scorer in LuceneSourceOperator.

```
at __randomizedtesting.SeedInfo.seed([C9F07BEF4F51E1]:0)
at org.apache.lucene.tests.index.AssertingLeafReader.assertThread(AssertingLeafReader.java:67)
at org.apache.lucene.tests.index.AssertingLeafReader$AssertingNumericDocValues.docID(AssertingLeafReader.java:689)
at org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:246)
at org.elasticsearch.compute.lucene.LuceneSourceOperator.getOutput(LuceneSourceOperator.java:119)
at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:179)
at org.elasticsearch.compute.operator.Driver.run(Driver.java:131)
at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:276)
at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
at java.base/java.lang.Thread.run(Thread.java:1623)
```

Most of our unit tests execute the `run` method of Driver which executes
operator uninterruptibly in a single thread. However, we should simulate
the yielding and rescheduling behavior when possible. This PR also
removes the run method forcing tests using the async execution of Driver
with a randomized max_iterations.

elasticsearchmachine committed:
Merge pull request ESQL-1428 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1433 from elastic/main

🤖 ESQL: Merge upstream

dnhatn committed:
Pass the actual config when generating layout (ESQL-1427)

We should pass the actual config instead of creating a dummy one when
generating the page layout of a remote exchange source.

Closes ESQL-1420

dnhatn committed:
Support multiple segments in enrich lookup (ESQL-1398)

While in general enrich indices have a single segment, this requirement
is not always guaranteed. This PR ensures that ESQL enrich lookup should
still be able to function with multiple segments.

elasticsearchmachine committed:
Merge pull request ESQL-1435 from elastic/main

🤖 ESQL: Merge upstream

dnhatn committed:
Adapt transport version change upstream

elasticsearchmachine committed:
Merge pull request ESQL-1436 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1437 from elastic/main

🤖 ESQL: Merge upstream

ChrisHegarty committed:
Use big array wrapped vectors in intermediate state of primitive aggs (ESQL-1426)

Use big array wrapped vectors in intermediate state of primitive aggs.

bpintea committed:
Declare metadata fields part of the `from` command (ESQL-1432)

This moves the declaration of the metadata fields from the dedicated
function (`metadata()`, which is dropped) to the index source command:
`from idx [metadata _index, _version]`.

Once these metadata fields are declared, they can be used in the query
just like the other field attributes, in evaluation, filtering, sorting,
grouping. They are also implicitly part of the `from` output, along with
- and past - the other index fields.
`from idx [metadata _version] | eval v = _version + 1 | where _version >
1`

Closes https://github.com/elastic/elasticsearch-internal/issues/1397.

craigtaverner committed:
Change pow() behaviour with unsigned long to return long

The original design was that floating types returned floating types, and non-floating returned non-floating, similar to the behaviour of MQ-SQL. However, the unsigned long support deviated from this by returning double. This change returns to the original behaviour.

One negative of this is that we can no longer return Infinity for cases of numerical overruns.

ChrisHegarty committed:
Remove/Revert new unused bits in BitArray (ESQL-1434)

This commit removes / reverts new changes in BitArray that are no longer
used.

craigtaverner committed:
Improve documentation for pow function and refined type rules

elasticsearchmachine committed:
Merge pull request ESQL-1443 from elastic/main

🤖 ESQL: Merge upstream

ChrisHegarty committed:
Revert whitespace

ChrisHegarty committed:
Revert to align with upstream

ChrisHegarty committed:
Disable javadocs for qa test projects (ESQL-1440)

This commit disables javadocs for qa test projects, since they are not
necessary or interesting, and cause unnecessary warning noise in the
build log output.

elasticsearchmachine committed:
Merge pull request ESQL-1444 from elastic/main

🤖 ESQL: Merge upstream

ChrisHegarty committed:
Revert Use big array wrapped vectors in intermediate state of primiti… (ESQL-1446)

elasticsearchmachine committed:
Merge pull request ESQL-1449 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1452 from elastic/main

🤖 ESQL: Merge upstream

ChrisHegarty committed:
Wrapper around TextDocValuesField that provides access to the SortedBinaryDocValues (ESQL-1453)

This commit adds a local wrapper around TextDocValuesField that provides
access to the SortedBinaryDocValues.

This avoids the need to more generally break encapsulation of
BaseKeywordDocValuesField.

dnhatn committed:
Fix needsInput of Limit operator (ESQL-1451)

The Limit operator can silently drop input pages when the downstream
operators are blocked. This occurs because its 'needsInput' method does
not consider the outstanding page.

Closes ESQL-1442
Closes ESQL-1431
Closes ESQL-1430

This bug was introduced in ESQL-1415.

dnhatn committed:
Fix offset in grouping stats test (ESQL-1450)

Some of grouping stats tests are failing. I think this is because the
chunking offset in these tests are incorrect.

ChrisHegarty committed:
Revert now unnecessary change to AttributeSet (ESQL-1454)

This was needed at one point to support different set implementations,
but is no longer required. Reverting.

elasticsearchmachine committed:
Merge pull request ESQL-1455 from elastic/main

🤖 ESQL: Merge upstream

dnhatn committed:
Increase driver iteration time limit (ESQL-1445)

Today, we pause the execution of a Driver and yield the executing thread
to another Driver in the following cases:

1. When the downstream or upstream pipeline is not fast enough. For
example, if the coordinator hasn't proceeded with all the output pages
from the data-node, the data-nodes can yield their executing threads.
This is done via the buffer of the exchanges.
2. When the number of iterations exceeds the limit (defaults to 10,000).
3. When the execution time exceeds the limit (defaults to 200ms).

I don't see the benefits of the second and third scenarios except when
dealing with slow queries. Throttling slow queries can provide
opportunities for fast queries to complete quickly. However, this
increases memory usage, as more outstanding requests are executed
simultaneously, instead of being limited to the number of threads.
Rescheduling drivers can also be expensive, as it involves
re-initializing certain Lucene structures.

In this pull request, we increase the time limit of a Driver iteration
from 200ms to 5 minutes.

elasticsearchmachine committed:
Merge pull request ESQL-1456 from elastic/main

🤖 ESQL: Merge upstream

luigidellaquila committed:
Add date_extract function (ESQL-1346)

elasticsearchmachine committed:
Merge pull request ESQL-1461 from elastic/main

🤖 ESQL: Merge upstream

craigtaverner committed:
Numerical overflow should result in `null` and a warning

To implement this we:

* Cast both arguments to double
* Perform integer and long validation on the double results before casting back to integer or long
* Perform a special case validation for exponent==1
* Any validation failures result in ArithmeticException, which is caught and added to warnings

elasticsearchmachine committed:
Merge pull request ESQL-1466 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1469 from elastic/main

🤖 ESQL: Merge upstream

craigtaverner committed:
Fixed failing tests after change to types rules

craigtaverner committed:
Fix flaky test, warnings were not always present

The fact that warnings were not always there is worrying.
I would expect this to be deterministic.
One theory is that the `limit 4` has a back-propagating effect, if the results are streamed.
This requires also that the chunk size is small enough that specific chunks do not include the value
causing the overflow, and chunk ordering and merging is such that that chunk is sometimes not
considered in time for the limit to cause the execution to complete.

craigtaverner committed:
Update docs/reference/esql/functions/pow.asciidoc

Co-authored-by: Bogdan Pintea <pintea@mailbox.org>

craigtaverner committed:
Code review cleanup

nik9000 committed:
Cut the output of LuceneTopNSource into many pages (ESQL-1429)

This cuts the output of `LuceneTopNSourceOperator` into pages so we can
run it on large numbers of documents.

elasticsearchmachine committed:
Merge pull request ESQL-1470 from elastic/main

🤖 ESQL: Merge upstream

dnhatn committed:
Allow filtering null on enrich fields (ESQL-1467)

The current resolver assumes that the enrich fields are not nullable;
however, they can be. This leads to a bug where we optimize away the
`is_null` or not `is_null` filters on enrich fields and return
unexpected results.

elasticsearchmachine committed:
Merge pull request ESQL-1471 from elastic/main

🤖 ESQL: Merge upstream

bpintea committed:
Add `_query` endpoint as `_esql` replacement (ESQL-1463)

This adds a new ES|QL endpoint, `_query`, to replace the now deprecated
`_esql`. The latter is still kept for a while, emitting a deprecation
warning.

Fixes ESQL-1379.

bpintea committed:
Change `RENAME`'s syntax from using `=` to `AS` (ESQL-1462)

This changes the `RENAME` syntax from `RENAME new = old` to `RENAME old
AS new`.

Fixes ESQL-1447.

bpintea committed:
Fix `In` folding with BytesRef types (ESQL-1457)

This fixes `In`'s folding in case of BytesRef-based types (Keyword/Text,
Version, IP). It now calls InProcessor directly, instead of using QL's
`In`, which does an implicit conversion, since this conversion cannot be
done succesfully, since QL doesn't support BytesRef (and even if it did,
QL's converters don't use associated source type, but try to derive it
from the received object).
This is possible since ESQL's `In` doesn't support implicit conversions
(except for non-unsigned_long numerics).

Fixes ESQL-1424.

not-napoleon committed:
Refactor function test cases  (ESQL-1405)

After working with AbstractFunctionTestCase a bit, I saw some
opportunities for improvement. This PR reduces the number of methods
function tests need to override, and provides a path towards generating
a matrix of test data and test paths.

craigtaverner committed:
Fix flaky tests with warnings checks

elasticsearchmachine committed:
Merge pull request ESQL-1475 from elastic/main

🤖 ESQL: Merge upstream

craigtaverner committed:
Removed system.out calls

nik9000 committed:
Fix error in csv tests (ESQL-1474)

Surprise! There was an off by one error in the CSV tests which made it
so the tests would pass if we specified exactly one fewer row than was
returned. More than one fewer row would fail. This corrects the error
and tries to make the code a little more "normal java" with the index
variable of the `for` loop not escaping. Escaping index variables have a
defined behavior, but I totally didn't see the off by one error until I
was testing something else. But there it was.

Anyway! This fixes the error and fixes the tests.

elasticsearchmachine committed:
Merge pull request ESQL-1480 from elastic/main

🤖 ESQL: Merge upstream

bpintea committed:
Support IN 3-valued logic (ESQL-1423)

This adds support for 3VL to IN:
- if there's a `null` involved in the comparisons and there's no match,
`null` is returned (instead of `false`);
- if there's a `null` involved, but there is a match, `true` is
returned;
- if there's no `null` involved, `true`/`false` is returned (as before).

The change also corrects the implementation, now evaluating each
equality just once. (The "per-row" iteration, allowing the evaluations
of each equality to stop at the first `true` is anyways no longer
possible due to the 3VL need to detect NULLs.)

Fixes ESQL-1232

craigtaverner committed:
Merge pull request ESQL-1439 from craigtaverner/esql/pow_unsigned_long

pow unsigned long

luigidellaquila committed:
Enable support for TEXT fields in GROK and DISSECT (ESQL-1482)

elasticsearchmachine committed:
Merge pull request ESQL-1485 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1487 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1489 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Size pages based on loaded columns (ESQL-1403)

This sizes pages produced by operators based on an estimate of the
number of bytes that'll be added to the page after it's been emitted.

At this point it only really works properly for the
`LuceneSourceOperator`, but that's pretty useful!

The `LuceneTopNSourceOperator` doesn't yet have code to cut the output
of a topn collected from lucene into multiple pages, so it ignores the
calculated value. We'll get to that in a follow up.

We feed the right value into aggregations but ungrouped aggregations
ignore it because they only ever emit one row. Grouped aggregations
don't yet have any code to cut their output into multiple pages.

TopN *does* have code to cut the output into multiple pages but the
estimates passed to it are kind of hacky. A proper estimate of TopN
would account for the size of rows flowing into it, but I never wrote
code for that. The thing is - TopN doesn't have to estimate incoming row
size - it can measure each row as it builds it and use the esimate we're
building now as an estimate of extra bytes that'll be added. Which is
what it is! But that code also needs to be written.

Relates to https://github.com/elastic/elasticsearch-internal/issues/1385

elasticsearchmachine committed:
Merge pull request ESQL-1490 from elastic/main

🤖 ESQL: Merge upstream

luigidellaquila committed:
Add support for Locale to ESQL (ESQL-1488)

elasticsearchmachine committed:
Merge pull request ESQL-1492 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1500 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1501 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1502 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Stop accidentally making tiny pages (ESQL-1506)

In ESQL-1403 I accidentally caused us to create *tiny* pages - 10 documents.
It's just using a `min` instead of a `max`. Ooops! And the tests didn't
notice. But the performance tests did!

This flips us back to making normal sized pages and adds a test that
should catch this. It's a unit test which isn't ideal, but other
mechanisms aren't really easy to get at right now and a unit test is
useful here anyway.

elasticsearchmachine committed:
Merge pull request ESQL-1510 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Support `auto_bucket` for numeric fields (ESQL-1494)

This adds support for numeric fields to `auto_bucket` and adds a new
`floor` function to round numeric down to the nearest integer. That
function is exposed because it's probably useful. I added it in this PR
because `auto_bucket` uses it as an implementation detail as well.

elasticsearchmachine committed:
Merge pull request ESQL-1512 from elastic/main

🤖 ESQL: Merge upstream

luigidellaquila committed:
Add support for multi-value input to DISSECT command (ESQL-1484)

nik9000 committed:
Change how grouping aggs track null (ESQL-1328)

This changes how grouping aggs track `null` to fix bugs in the old
tracking caused by `null`s arriving after non-null values and things
like that. Now there are two "modes" from the `X-ArrayState` classes:
1. A mode where we do not track null and rely on the `selected` list of
buckets produced by the `BlockHash`. This mode is appropriate right up
until you see the first `null` value.
2. A mode where we do track which values are null and can reply with
`null` even for values `BlockHash` selects if we've only ever added
`null` values to that slot. This is appropriate for all data, but less
efficient so we only transition to it when we receive our first `Block`
containing `null` values.

The transition is *interesting* because we need to know which values
aren't `null`. Luckily, `BlockHash` has that information. Usually it's
just "everything in this range" but for `boolean` values it isn't. And,
I expect, for ordinals it won't be either. So we just ask `BlockHash` at
the moment of the transition. That required changing the interface
around a little bit so we could ask *before* the group keys for the next
block was added.

bpintea committed:
Exact math (ESQL-1507)

This turns the current math ops eval implementations to work exactly,
i.e. return null on failures caused by over-/underflows or division by
zero and corresponding Warnings.

Fix ESQL-1459

elasticsearchmachine committed:
Merge pull request ESQL-1517 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Add trigonometric functions (ESQL-1513)

This adds `SIN`, `COS`, `TAN`, `SINH`, `COSH`, and `TANH` functions.

---------

Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>

nik9000 committed:
Add query to esql task (ESQL-1511)

This adds the `query` to the main ESQL task so you can see long running
queries. And adds some docs about it including an example of cancelling
a query.

---------

Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>

csoulios committed:
Add support for `_id` fields (ESQL-1505)

Query example to retrieve `_id` 
```
from hosts [metadata  _id] | keep _id
```

elasticsearchmachine committed:
Merge pull request ESQL-1522 from elastic/main

🤖 ESQL: Merge upstream

nik9000 committed:
Fix error loading sorted docs from empty (ESQL-1514)

This fixes an error when loading docs from `LuceneTopNSourceOperator`
when the index shard is empty. We'd `NullPointerException` because we
assumed at least one "leaf" was present in the `IndexReader`.

Closes ESQL-1508

nik9000 committed:
Add remaining trigonometric functions (ESQL-1518)

Adds the remaining trigonomentric functions, `ACOS`, `ASIN`, `ATAN`, and
`ATAN2`.

---------

Co-authored-by: Bogdan Pintea <pintea@mailbox.org>

elasticsearchmachine committed:
Merge pull request ESQL-1525 from elastic/main

🤖 ESQL: Merge upstream

costin committed:
Infer layout between plans instead of storing it (ESQL-1519)

The big change in this PR is exposing the aggregation intermediate state
 in the plan. Thus the output of the data node plan and input of the
 coordinator plan end up the same and thus the layout can be inferred
 without having to be memorized (or determined from the subplan).
To make that happen, the page traveling between plans must have the same
order and thus are aligned as there's no map to hold any out of order
information.

Furthermore in case of aggregations, since the NameIds are not the same
 across plans, a 'decorating' layout is used that looks at the call
 order instead of the actual arguments passed to it. This clearly needs
 improving.

nik9000 committed:
Add `to_degrees` and `to_radians` functions (ESQL-1496)

This adds the `to_degrees` and `to_radians` functions. It uses the
"convert" function framework because that just felt right - these
convert between radians and degrees after all.

costin committed:
Pick upstream change

nik9000 committed:
Fix ESQL compilation

Fix the compilation after merging #98131 which removed a constructor
that we were using. The old constructor used the transport thread to
execute the response. I checked all of the cases where the compilation
failed and all of those places looked appropriate to run on the
transport worker.

elasticsearchmachine committed:
Merge pull request ESQL-1527 from elastic/main

🤖 ESQL: Merge upstream

bpintea committed:
Switch `mv_sum` to exact math (ESQL-1523)

This switches the `mv_sum` function to perform exact math: return null
and issue Warning headers on overflow.

Following ESQL-1507.

not-napoleon committed:
Parameterized testing for ESQL functional tests (ESQL-1497)

Convert the function tests to use parameters. I converted some test cases, but not all of them (will add more in follow up PRs)

elasticsearchmachine committed:
Merge pull request ESQL-1530 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1536 from elastic/main

🤖 ESQL: Merge upstream

luigidellaquila committed:
Remove unused import from SQL build.gradle (ESQL-1537)

bpintea committed:
Switch `mv_avg` to double summation for integral types (ESQL-1531)

This switches the `mv_avg` to perform double summation also for the
integer types. This allows correct returns also for the cases when the
summation would otherwise cause an overflow in the input (integral)
type.

The summation is compensated, just like for the double input case.

bpintea committed:
Optimize `In` when matching value is foldable (ESQL-1468)

This adds a logical optimisation rule affecting IN that has a foldable
value to compare against a list, separating all the other foldable
values in the values list into a different IN that can then be folded
entirely.

This change also turns an IN that only compares against one element into
an equality.

These changes allow to further simplify the query, recombine the
conditions into a simpler one or enable push down.

Resolves ESQL-1391.

luigidellaquila committed:
Fix Locale management in ESQL query request (ESQL-1538)

nik9000 committed:
Shrink "huge" combinatorial tests a little (ESQL-1476)

The "huge" combinatorial tests can take a few minutes to complete
sometimes and that's just a real pain. This shrinks them somewhat so on
average they'll be twice as fast - at worst maybe a two minutes.

nik9000 committed:
Group on `null`s in single fields  (ESQL-1479)

This adds support for grouping on `null` values in the same way that
databases do - the `null` value is a grouping key. Previously we'd skip
all rows with a `null` value in the grouping key. This fixes that for
all the single valued fields. For example, you'd see:
```
count(a) | b
       1 | null
       2 | 1
       3 | 2
```

instead of what we'd previously do:
```
count(a) | b
       2 | 1
       3 | 2
```

This only fixes the grouping when you group on a single field.
Multi-field groupings don't yet work.

Relates to https://github.com/elastic/elasticsearch-internal/issues/1327

elasticsearchmachine committed:
Merge pull request ESQL-1542 from elastic/main

🤖 ESQL: Merge upstream

costin committed:
Locally optimize missing fields (ESQL-1532)

Refine local plans based on the available local information. 
Introduce optimization that checks if any of the fields 
 defined are locally and replacing them with null.

elasticsearchmachine committed:
Merge pull request ESQL-1543 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1545 from elastic/main

🤖 ESQL: Merge upstream

not-napoleon committed:
Parameterized testing follow up (ESQL-1540)

Addresses a few follow ups from
https://github.com/elastic/elasticsearch-internal/pull/1497.
Specifically, drops the need to specify `Source` on all test cases
(instead setting `Source.EMPTY` always, which in practice was what we
were doing anyway) and moving the expected return type into the test
case.

It does not remove the abstract method for checking the return type,
since that requires some further changes in other looping tests.
Addressing those will be the next PR.

elasticsearchmachine committed:
Merge pull request ESQL-1546 from elastic/main

🤖 ESQL: Merge upstream

elasticsearchmachine committed:
Merge pull request ESQL-1551 from elastic/main

🤖 ESQL: Merge upstream

abdonpijpelink committed:
Add tech preview admonition

abdonpijpelink committed:
Update supported types list

abdonpijpelink committed:
Update types list

elasticsearchmachine committed:
Merge pull request ESQL-1553 from elastic/main

🤖 ESQL: Merge upstream

abdonpijpelink committed:
Merge pull request ESQL-1552 from abdonpijpelink/docs-tech-preview-admonition

[DOCS] Mark ES|QL as 'tech preview' and update supported types list

costin committed:
Blog examples (ESQL-1544)

ChrisHegarty committed:
Merge branch 'main' into feature/esql

ChrisHegarty committed:
Remove docs workaround (#98346)

This commit removes a temporary workaround that was added for the doc root.

ChrisHegarty committed:
Test fix: add ESQL usage as a non-operator action (#98347)

We did not define if esql actions are operator only or non operator action which resulted in the following failure:

java.lang.AssertionError: Actions are neither operator-only nor non-operator: [[
cluster:monitor/xpack/esql/stats/dist, cluster:monitor/xpack/info/esql,
cluster:monitor/xpack/usage/esql, indices:data/read/esql]].
An action should be declared to be either operator-only in 
[org.elasticsearch.xpack.security.operator.DefaultOperatorOnlyRegistry] or non-operator in 
[org.elasticsearch.xpack.security.operator.Constants]

luigidellaquila committed:
ESQL: Add Multi-Value support to GROK operator (#98316)

bpintea committed:
ESQL: Check the timespan value against the allowed range (#98319)

This checks the value provided for the timespan literal against the type
range of the underpinning constructor.
It also avoids a CCE thrown that had occurred when input value
exceeded int's range.

ChrisHegarty committed:
Merge upstream

luigidellaquila committed:
ESQL: Add support for negative numbers in multivalues for ROW operator (#98359)

bpintea committed:
ESQL: Pushdown metadata attributes (#98355)

This allows to push down those metadata attributes that can be pushed
down.
Out of the currently supported three, the only one pushed down is
_index. This only supports wildcard and terms queries.

bpintea committed:
Verify the constant types assigned in a row (#98403)

This adds a verification for the constant types assigned directly into
a ROW source command.
Types like Duration and Period cannot be used, as they are not
representable in a block.

ChrisHegarty committed:
Merge upstream

ChrisHegarty committed:
Add unavailable EsqlFeatureSetUsage creation support (#98438)

This commit adds a means to create an EsqlFeatureSetUsage that is unavailable and not enabled. This required to be able to mock out alternative / dummy implementations of usage / info transport actions where the ESQL module is not present, thus returns empty / unavailable feature usage stats, etc.

This is a similar to other not-present xpack features in stateless, e.g. ILM.

ChrisHegarty committed:
Remove space in dissect feature list

ChrisHegarty commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -266,6 +266,7 @@ private static boolean shouldAuthorizeIndexActionNameOnly(String action, Transpo
             case "indices:data/write/reindex":
             case "indices:data/read/sql":
             case "indices:data/read/sql/translate":
+            case "indices:data/read/esql":
```
@ywangd This is a big PR, and I do not expect a full review, but I do want to call out this change.

ywangd replied:
Thanks for the ping. This particular change here looks right to me after briefly reading through relevant code around this change. Similar to the SQL API, the indices that ES|QL accesses is not known when the request is initially authorized. Therefore we only authorize it by the action names without looking into the indices. The index authorization will be performed when the ES|QL services send out shard level request for data access. These shard level requests are

1. [DataNodeRequest](https://github.com/elastic/elasticsearch/blob/2e4440ff1c0decf79f8a40688566411f7d518aa7/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequest.java#L32) - for data accessing and computation
2. [LookupRequest](https://github.com/elastic/elasticsearch/blob/2e4440ff1c0decf79f8a40688566411f7d518aa7/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichLookupService.java#L268) - for data encrichment

Both requests implement `IndicesRequest`, have corresponding transportHandlers, and are issued with `transportService.sendChildRequest`. Therefore they go through the existing index authorization process and properly authorized. 

_One question_ I do have is why we chose to [execute the `LookupRequest` with the `enrich` origin](https://github.com/elastic/elasticsearch/blob/2e4440ff1c0decf79f8a40688566411f7d518aa7/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichLookupService.java#L141C2-L141C2). It means the following query works when the user has access to only the `playlist` index without access to either the `songs` enrich policy or its associated index. 

```
FROM playlist | ENRICH songs ON song_id
```

It feels a bit weird to me that we have no control on who can use what enrich process. That said, we don't have a good story around enrich and ingest pipeline either, i.e. we have no control over who can use what pipeline where a pipeline can contain an enrich processor. There have been quite a few complaints from users and we have an [open issue](https://github.com/elastic/elasticsearch/issues/57586) for it. I wonder whether we should *not* repeat the problem here with ES|QL. On the surface, it might be easier to tackle compared to ingest pipelines since ES|QL is read-only and has no "bulk" operations. What do you think?

ChrisHegarty replied:
Thank you so much @ywangd . This is a great question / suggestion, I filed the following issue to track it   #98482

dnhatn replied:
@ywangd The primary reason we perform the enrich lookup with the enrich origin is to prevent the warning that arises when accessing enrich indices, which are system indices. An alternative is to override `allowSystemIndexAccessByDefault` in the rest layer, but I felt that it might be too broad. I'm happy to iterate on it.

ywangd replied:
I posted my reply on the issue (#98482) raised by Chris to have future discussions in one place and avoid disturbing all people involved in this huge PR.

ChrisHegarty committed:
Merge upstream

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

nik9000 committed:
ESQL: Support `null` group key in more cases (#98507)

This adds support for null group keys in all groupings that are
backed by the PackedValuesBlockHash. This means the only groupings
that don't support null keys at the moment are long/long pairs,
and long/bytes_ref pairs. Those are coming next.

nik9000 committed:
ESQL: Optimize some `MV_` functions on sorted (#98515)

When multivalued fields are loaded from lucene they are in sorted order
but we weren't taking advantage of that fact. Now we are! It's much
faster, even for fast operations like `mv_min`

```
     (operation)  Mode  Cnt  Score   Error  Units
          mv_min  avgt    7  3.820 ± 0.070  ns/op
mv_min_ascending  avgt    7  1.979 ± 0.130  ns/op
```

We still have code to run in non-sorted mode because conversion functions
and a few other things don't load in sorted order.

I've also ported expanded the parameterized tests for the `MV_` functions
because, well, I needed to expand them at least a little to test this
change. And I just kept going and improved as many tests as I could.

ChrisHegarty committed:
Merge upstream

elasticsearchmachine commented:
Hi @ChrisHegarty, I've created a changelog YAML for you.

ChrisHegarty committed:
Update docs/changelog/98309.yaml

ChrisHegarty committed:
Post-merge fix - update TransportResponseHandler.Empty::handleResponse

ChrisHegarty committed:
Merge upstream

ChrisHegarty committed:
Fix changelog area

ChrisHegarty committed:
Post-merge fix - update TransportResponseHandler.Empty::handleResponse (again)

ChrisHegarty committed:
Declare and assign TransportVersion for EsqlFeatureSetUsage (#98525)

This commit adds a new TransportVersion for EsqlFeatureSetUsage.

ChrisHegarty committed:
Merge upstream

kkrik-es committed:
`Sqrt` function for ESQL (#98449)

* Sqrt function for ESQL

Introduces a unary scalar function for square root, which is a thin
wrapper over the Java.Math implementation.

* Fix area for ESQL integration changelog.

* Restore changelog.

* Restore area in changelog.

ChrisHegarty committed:
Merge upstream

alex-spies committed:
ESQL: Enable expressions with arithmetic negation (#98352)

Enable unary arithmetic negations in expressions, like eval x = -y.
Support integer, long and double arguments.

Changes:
* Add mapper for Neg expressions.
* Add tests.
* Disallow negating unsigned longs during query verification.

ChrisHegarty committed:
Merge upstream

astefan committed:
ESQL: replace the is_null function with IS NULL and IS NOT NULL predicates (#98412)

costin committed:
Prune unused columns from the query (#98462)

Remove unused aggregations or eval fields that are not needed in the
 final result.
For example the query:

from employees
| stats c = count(salary), min = min(salary)
| eval x = emp_no
| keep c

is now optimized as:

from employees
| stats c = count(salary)
| keep c

since neither the min or nor the x fields are actually needed.

ChrisHegarty committed:
Merge upstream

ChrisHegarty commented:
@elasticsearchmachine rerun elasticsearch-ci/bwc

elasticsearchmachine commented:
@ChrisHegarty according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is labelled `release highlight` but the changelog has no `highlight` section
